### PR TITLE
- Load only available ports to the settings dialog

### PR DESCRIPTION
**Pull request by martonmiklos**
- Allow setting the first and last scanned address per port in the
settings dialog
- Get rid from the rapidjson based settings and use QSettings
- Introduce protocolcombobox class for protocol selection

### CODE REVIEW

**Comment by emericg** [comment]
Seems great. I'll test it tomorrow before merging.
The travis CI error is just a missing package, but the AppVeyor seems like a real bug. Is it related to the minimum Qt version problem you were talking about?

**Comment by martonmiklos** [comment]
>  Is it related to the minimum Qt version problem you were talking about?

Yeah, it is. It looks like the designer saved it before commit again. 
Let me fix it., and I will look after how can we install serialport on travis as well.

**Comment by martonmiklos** [comment]
Cool looks like we are passing on travis and appveyor now.

**Comment by emericg** [comment]
Hi,
Great to see that the tests are passing.

I tried the patches today, with some success. The ports seemed detected on linux and macOS (but came doubled, I'll fix that later if you don't have access to a mac don't worry). Settings are correctly saved.
However I had a couple of problems. First, the protocol comboboxes are empty on first runs, until we set a value manually.
I'm using XL-320 but couldn't set and use my usual settings (v2, 1Mbps), I'm not sure why. I'll have a closer look tomorrow morning.

Regarding the feature, it seems mostly allright but the autodetected ports shouldn't be used by default (left combobox shouldn't be ticked) so that the protocol/speed auto-detection is always run first, until the user set values manually. Also, why are fonts in bold? ^^
One more remark, the id range should go up to 254. It's only when using some adapters that the id 254 is reserved.

**Comment by martonmiklos** [comment]
> (but came doubled, I'll fix that later if you don't have access to a mac don't worry). 

Yeah one of my colleagues using macOS and he said he identified the same behavior. He just ignored the port startin with "cu."

> First, the protocol comboboxes are empty on first runs, until we set a value manually.

 I set Dynamixel V1 as default (this was the default before the patch as well)

> Regarding the feature, it seems mostly allright but the autodetected ports shouldn't be used by default (left combobox shouldn't be ticked) so that the protocol/speed auto-detection is always run first, until the user set values manually. 

Fixed.

> the id range should go up to 254

Fixed

> Also, why are fonts in bold? ^^

Inherited from the groupbox font settings.  I would prefer styling via CSS. If you agree I will convert the font settings to stylesheets.

Also I am thinking to convert the baud rate spinbox to combobox. QSerialPort has a standardBaudRates member which lists the available baudrates on the specific target.

**Comment by emericg** [comment]
Almost there ^^
When saving settings, range seems to be taken into account but neither protocol nor speed are.

> Inherited from the groupbox font settings.

I remember having similar troubles with that...
Also I noticed the "serial port" groupbox now includes the "device controller" groupbox, probably a manipulation mistake.
You can change the little text (the 'preconfigure up to 4 ports') that explain the feature so it's more up to date with the feature changes.

> I would prefer styling via CSS. If you agree I will convert the font settings to stylesheets.

Sure why not. I'm using CSS styling in other places.

> Also I am thinking to convert the baud rate spinbox to combobox. QSerialPort has a standardBaudRates member which lists the available baudrates on the specific target.

I'm not sure about that. Dynamixel seems to accept every possible values for speed, way more permissive then the serial ports seems to be. So I fear that if we accept only a subset of possible speeds, some users may be stuck with inaccessible servos?
I guess if we only accept some particular speed in the settings but that we can still use manual speed in the main UI and advanced scanner it would be OK.

**Comment by martonmiklos** [comment]
> I'm not sure about that. Dynamixel seems to accept every possible values for speed, way more permissive then the serial ports seems to be. So I fear that if we accept only a subset of possible speeds, some users may be stuck with inaccessible servos?
> I guess if we only accept some particular speed in the settings but that we can still use manual speed in the main UI and advanced scanner it would be OK.

Makes sense, I have a better idea: the QComboBox can be made editable. I think that would be the best fit in this case. I will check and address the mentioned issues soon. 

- CSS styling fixed (only the groupboxes titles are bold now)
- I have moved the Device controller groupbox to the old place, it was mistakenly moved there

**Comment by martonmiklos** [comment]
>  It's only when using some adapters that the id 254 is reserved.

Are you sure about this? 254 seems to be the broadcast address for Dynamixel and Herkulex as well.
For scanning we should never need to include the broadcast address right?

I have made the following changes meantime: 

- Introduced an editable, validate combobox for the baud rates
- The settings from the Settings dialog now propagated properly to the port widgets of the Mainwindow.

BTW. what should be the save strategy for the state of the serial port settings on the main window? At the moment if the user sets the port parameters only in the main window (with manual port mode), the settings  got lost on quit which is kind of awkward solution. I would save the values of those widgets on the program close. What do you think?

**Comment by emericg** [comment]
> Are you sure about this? 254 seems to be the broadcast address for Dynamixel and Herkulex as well.
For scanning we should never need to include the broadcast address right?

Damn you are right, I was thinking of id 253 that is reserved when using usb2ax.

I'll test the latest modifications tomorrow and merge this if it is working. Further behavior change can be a separate pull request.
By the way force pushing your changes is not necessary when you have only one patch and a well defined feature. You can add commits to the pull request, that would make changes easier to track, and I can squash the commits in a single one when merging.
In case of multiple patches in one pull request, force pushing like that would keep it clean and tidy yes.

Regarding behavior, I don't know. Automatically saving known port settings might be a good idea yes, I can't think of situation where it would be worst than current behavior anyway. If a user is juggling with several ports / servos series, settings saved or not he probably would have to adjust them manually anyway.
But the thing is, how do you know when the settings, even if set manually, are the right ones? They should be saved only if servos are indeed detected right?
So if the auto-detection is still run when a port is first found and settings are saved only when servos are detected, it would be pretty neat.

**Comment by martonmiklos** [comment]
Alright, I have just pushed one more feature before I read this:
The "rescan serial ports action" removes/adds the ports the settings window as well. 

I have managed to ran some issues with this:  the QGridLayout row count does not decreasing even if you removed all items in the last a row. But anyway I have workarounded it.

**Comment by emericg** [comment]
> The "rescan serial ports action" removes/adds the ports the settings window as well.

But what's the point in saving settings if they are removed every time the adapter is unplugged? That was the original idea. I use a couple of adapters on a couple of computers (or at least I used to ^^)

To talk a little bit more about behavior, the ports:
- should be saved on file only when they are manually ticked or servos are detected (for instance on macs, the bluetooth modem always appear as a serial device, chances are users will not ever want to save these)
- saved ports should be persistant across unplugging
- saved ports should always appear on the setting UI
- there should be a delete button?

does that makes sense?

**Comment by martonmiklos** [comment]
> But what's the point in saving settings if they are removed every time the adapter is unplugged? That was the original idea. I use a couple of adapters on a couple of computers (or at least I used to ^^)

Let me rephrase the 'The "rescan serial ports action" removes/adds the ports the settings window as well.' because yeah it was a bit confusing:
If I hit the rescan button after a new serial port added to the system it is going to be listed on the main window, and now in the settings window too. And the removal works the opposite way in the settings window too when unplugging an interface.

File: SmartServoGui/src/widgetSerialScan.cpp [hunk 1]
```diff
@@ -163,34 +181,31 @@ int widgetSerialScan::getCurrentDeviceClass(int index)
...
     {
-    case 1:
-    case 11:
-        if (ui->comboBox_protocol->currentIndex() == 0)
+    case Manual:
+    case Saved:
+        if (ui->comboBox_protocol->currentProtocol() == ServoProtocol::PROTOCOL_DXLv1)
             deviceClass = SERVO_MX;
-        else if (ui->comboBox_protocol->currentIndex() == 1)
+        else if (ui->comboBox_protocol->currentProtocol() == ServoProtocol::PROTOCOL_DXLv1)
```
Comment (emericg) [inline SmartServoGui/src/widgetSerialScan.cpp]: should be v2! took me a while to figure that out :p

**Comment by emericg** [comment]
It's getting there... it's getting there...
Beside the bug mentioned in comment, I think the autoscan was 'on' by default and now it's off, that's important.
Can you set three sections in the setting file like before? It would be a bit more clean. There was gui, controller and serialport I think.
Other than that, the saved but not connected port are not still not shown.

File: SmartServoGui/src/widgetSerialScan.cpp (hunk 1 above)
Comment (martonmiklos) [reply SmartServoGui/src/widgetSerialScan.cpp]: Sorry about that, fixing it now.

**Comment by emericg** [comment]
file lock was probably enabled by default too.

**Comment by martonmiklos** [comment]
I have just realized that the non-port related settings have not been loaded to the UI. Just pushed the fix.

**Comment by martonmiklos** [comment]
> Other than that, the saved but not connected port are not still not shown.

Yes implementing this behavior mentioned above is still a pending task. But what is the added value to show ports not present but saved?

**Comment by emericg** [comment]
> But what is the added value to show ports not present but saved?

You can modify them even if they are not connected at the moment. One thing that I can think of would be removing settings that make the app crash. It can happen ^^

Beside that I think we are good to go with this pull request. And even without, I can push it now if you want to work this out later (or I can do it later too).

**Comment by emericg** [comment]
By the way should I move forward and merge this PR or do you plant to work on it again?

**Comment by martonmiklos** [comment]
Hi, 

I apologise for the lack of feedback, I had to focus on other things, so it slightly got forgotten. Please merge it, we can get back to "implementing the editing of the settings of the not connected ports" later. 

Thanks again for your time and effort, it was a pleasure to work with you!

**Comment by emericg** [comment]
Allright this is merged!
Thank you for this very cool feature, and it's been a pleasure to work with you too ;)
